MIDAS/src/telemetry.cpp

```diff
@@ -0,0 +1,119 @@
+#include "telemetry.h"
+
+
+/**
+ * @brief This function maps an input value onto within a particular range into a fixed point value of a certin binary
+ * size
+ *
+ * @param val: number to map into target range, values outside of the range will be clamped
+ *
+ * @param range: range to map number into. For unsigned output, [0, range). For signed output [-range/2, range)
+ *
+ * @return fixed point value represing val mapped onto the target range
+ */
+template <typename T>
+T inv_convert_range(float val, float range) {
+    size_t numeric_range = (int64_t)std::numeric_limits<T>::max() - (int64_t)std::numeric_limits<T>::min() + 1;
+    float converted = val * (float)numeric_range / range;
+    return std::max(std::min((float)std::numeric_limits<T>::max(), converted), (float)std::numeric_limits<T>::min());
+}
+
+
+void Telemetry::transmit(RocketData& rocket_data) {
+    telemetry_command command { };
+    while (backend.read(&command)) {
+        handleCommand(command);
+    }
+
+    if (!isnan(set_frequency_to)) {
+        backend.setFrequency(set_frequency_to);
+        set_frequency_to = NAN;
+    }
+
+    TelemetryPacket packet = makePacket(rocket_data);
+    backend.send(packet);
+}
+
+
+/**
+ * @brief  This function handles commands sent from the ground station
+ * to TARS. The effects of this function depend on the command
+ * sent.
+ *
+ * @param cmd: struct containing information necessary to process
+ *             ground station command.
+ *
+ * @return void
+ */
+void Telemetry::handleCommand(const telemetry_command &cmd) {
+    /* Check if the security code is present and matches on ground and on the rocket */
+    if (cmd.verify != std::array<char, 6>{'A', 'Y', 'B', 'E', 'R', 'K'}) {
+        return;
+    }
+    /* Check if lasted command ID matched current command ID */
+    if (last_command_id == cmd.cmd_id) {
+        return;
+    }
+    last_command_id = (int16_t)cmd.cmd_id;
+
+    if (cmd.command == SET_FREQ) {
+        set_frequency_to = cmd.freq;
+    }
+
+    if (cmd.command == SET_CALLSIGN) {
+        memcpy(callsign, cmd.callsign, sizeof(cmd.callsign));
+        Serial.println("[DEBUG]: Got callsign");
+    }
+}
+
+
+TelemetryPacket Telemetry::makePacket(RocketData& data) {
```

Comment:
We also want continuity, pyro state, and fsm to be sent over telem. I see it defined within the packet struct but it's never filled